#166314469 Implement Book and Author model feature

henryade opened the pull request:
#### What does this PR do?
Adds Author and Book model

#### Description of Task to be completed?
Add Author model
Add Book model

#### How should this be manually tested?
- clone the repository and checkout to this branch
- run `npm install` to add new dependencies
- update/create the env file with your database details using the development node_env
- run `npm run db:roll` command
- check your database to see if tables are created
- you can run `npm run db:back` to remove tables 

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#166314469](https://www.pivotaltracker.com/story/show/166314469)

#### Screenshots (if appropriate)
N/A 

#### Questions:
N/A

Lundii reviewed (commented):
I follow your instruction. After running npm run db:roll, the tables where created but I got this error. Why am I getting it?
![image](https://user-images.githubusercontent.com/37886814/58600694-9000b880-827d-11e9-8816-b1bd4a8523d6.png)

Adekoreday reviewed (changes requested):
great job, you need to delete your migrations and seed files the task says to implement only the model
kindly work on your database associations in the book model a book should also have a user that has borrowed it.
So it needs to be associated with the user somehow

henryade commented:
> I follow your instruction. After running npm run db:roll, the tables where created but I got this error. Why am I getting it?
> ![image](https://user-images.githubusercontent.com/37886814/58600694-9000b880-827d-11e9-8816-b1bd4a8523d6.png)
I believe that error shows up because the index file is empty.the only way to remove the error is either to delete the index file or populate it. And I am not asked to do that.

henryade commented:
> great job, you need to delete your migrations and seed files the task says to implement only the model
> kindly work on your database associations in the book model a book should also have a user that has borrowed it.
> So it needs to be associated with the user somehow

From my pivotal tracker, my task is to create both the model and the migrations file. I spoke to the TTL about seeding it and he said I don't need to. And I followed the diagram explicitly, in creating the table so I will need to ask about the user Id before adding it. Maybe the association my be created as another table.  Thanks

Lundii commented:
> > I follow your instruction. After running npm run db:roll, the tables where created but I got this error. Why am I getting it?
> > ![image](https://user-images.githubusercontent.com/37886814/58600694-9000b880-827d-11e9-8816-b1bd4a8523d6.png)
> > I believe that error shows up because the index file is empty.the only way to remove the error is either to delete the index file or populate it. And I am not asked to do that.

I think the TTL said it can be deleted.

Ukhu commented on server/database/migrations/20190529204216-create-author.js:
```diff
@@ -0,0 +1,21 @@
+      fullname: {
+        type: Sequelize.STRING,
+        allowNull: false
+      }
```
Your Authors migration file should contain the 'createdAt' and 'upatedAt' fields that Sequelize provides by default. Don't remove them.

henryade replied:
> Your Authors migration file should contain the 'createdAt' and 'upatedAt' fields that Sequelize provides by default. Don't remove them.

I will add them

Ukhu commented on server/database/models/author.js:
```diff
@@ -0,0 +1,11 @@
+module.exports = (sequelize, DataTypes) => {
+  const Author = sequelize.define('Author', {
+    fullname: DataTypes.STRING
+  }, {});
```
There is no 'id' property in your Authors model. It was specified in the DB diagram so it should be here also.

henryade replied:
The 

> There is no 'id' property in your Authors model. It was specified in the DB diagram so it should be here also.

The id property is auto generated so I don't see any need adding it here

Ukhu replied:
For the sake of clarity, it's good that you add it there since it was explicitly specified in the db diagram.

henryade replied:
okay

henryade replied:
done

Ukhu commented on server/database/models/author.js:
```diff
@@ -0,0 +1,11 @@
+  }, {});
+  Author.associate = function(models) {
+    // associations can be defined here
+  };
```
Since you're creating the books model as well, the Authors model should be associated with it.

henryade replied:
> Since you're creating the books model as well, the Authors model should be associated with it.

It's the book model that's associated with the authors model from the digram.  The author doesn't have association with the book model

Ukhu replied:
Associations are a two-way thing. From the diagram, a book 'belongsTo' one author, while an author can have many books, so a 'hasMany' association ought to be there.

henryade replied:
so you are saying I should add a bookid to the author's table?

Ukhu replied:
No, the author already has an authorId in the book table.

henryade replied:
yeah so what do i link the new association to?
author (id) to book(authorId)?

Ukhu replied:
Yes

Ukhu reviewed (commented):
Please create seeder files for the author model and book model. Also correct your commit message. It shouldn't start with 'ft'.

henryade commented:
> > > I follow your instruction. After running npm run db:roll, the tables where created but I got this error. Why am I getting it?
> > > ![image](https://user-images.githubusercontent.com/37886814/58600694-9000b880-827d-11e9-8816-b1bd4a8523d6.png)
> > > I believe that error shows up because the index file is empty.the only way to remove the error is either to delete the index file or populate it. And I am not asked to do that.
> 
> I think the TTL said it can be deleted.

Yes will delete it as soon as the electricity is restored

henryade reviewed (commented):
> Please create seeder files for the author model and book model. Also correct your commit message. It shouldn't start with 'ft'.

I spoke to the TTL and he said I don't need to create seeders file. I will edit the commit.

Ukhu commented:
> > Please create seeder files for the author model and book model. Also correct your commit message. It shouldn't start with 'ft'.
> 
> I spoke to the TTL and he said I don't need to create seeders file. I will edit the commit.

Ok. You can leave it out if he asked you to

Lundii reviewed (dismissed):
Good

Ukhu commented on server/database/models/books.js:
```diff
@@ -0,0 +1,21 @@
+  }, {});
+  Books.associate = function(models) {
+    // associations can be defined here
+    Books.belongsTo(models.author, {
```
It should be 'models.Author'. The author should be capitalized here.

henryade replied:
the author.js file in models folder is in small letters.

Ukhu replied:
In your Author definition, you used capital letters. That is how it should be here also.

henryade replied:
thats the authors database and it autogenerated. this is referencing the file in the models folder

I think you need to change it, It must be in capital letter

Adekoreday replied:
you need to rebase this branch

Adekoreday replied:
it should be models.Authors

looking at your db the table name is Authors
![image](https://user-images.githubusercontent.com/44194416/58656548-647cdd00-8314-11e9-9a14-03adcd2f294e.png)

Adekoreday replied:
good job overall make the changes so i can approve

Ukhu replied:
> thats the authors database and it autogenerated. this is referencing the file in the models folder
> 

On my local, I tried it using 'models.Author' and the association worked, so I don't think it's the name of the file that matters here. So for readability sake, I suggest you use 'Author', but that's up to you since it works with 'author' as well. Every other thing looks fine. Rebase the branch so I can approve. Cheers 👍.

henryade replied:
done

Lundii reviewed (dismissed):
Good job

KvNGCzA commented on server/database/models/author.js:
```diff
@@ -0,0 +1,12 @@
+    fullname: DataTypes.STRING
+  }, {});
+  Author.associate = function(models) {
+    // associations can be defined here
```
There should be an association between the `Author` and the `Books` here

ad3ola committed:
166314469-ft(bookModel): implement book and author model feature
implement author model
implement book model
added foreign key constraint
[finishes #166314469]

JuwonAbiola reviewed (approved):
good work!
